bin/run-ansible.sh

```diff
@@ -156,73 +156,160 @@ for extra_var in $extra_vars; do
 done
 IFS=${old_IFS}
 
-github_ssh_public_key_path="${HOME}/.ssh/id_rsa.pub"
-github_ssh_private_key_path="${HOME}/.ssh/id_rsa"
-if [[ $extra_vars == *"pr_branches="* ]]; then
-    echo " -------------------------------------------------------------------------------------"
-    echo "Testing SSH connection to Github with ssh -oStrictHostKeyChecking=no -T git@github.com"
-    echo "Using SSH key from $github_ssh_private_key_path"
-    ssh_test=$(ssh -oStrictHostKeyChecking=no -T git@github.com 2>&1 &)
-    if [[ "$ssh_test" == *"You've successfully authenticated"* ]]; then
-        echo " ---------------------------------"
-        echo "Github SSH key successfully added!"
-        echo " ---------------------------------"
+is_repo_public() {
+  local user_slash_repo=$1
+  ERROR=$(curl -fsS "https://api.github.com/repos/${user_slash_repo}" 2>&1 >/dev/null)
+  if [ $? -eq 0 ]; then
+    printf '%s\n' "The GitHub repo ${user_slash_repo} exists." >&2
+    echo "true"
+  else
+    if [[ "${ERROR}" == *"error: 403"* ]]; then
+      echo "403"
     else
-        if [[ -z ${read_input} ]]; then
-            read_input="github_email"
+      printf '%s\n' "Error: no GitHub repo ${user_slash_repo} found." >&2
+      echo "false"
+    fi
+  fi
+}
+
+confirm() {
+	# call with a prompt string or use a default
+	read -r -p "${1:-[y/N]} " response
+	case "$response" in
+		[yY][eE][sS]|[yY])
+			echo "y"
+			;;
+		*)
+			echo "n"
+			;;
+	esac
+}
+
+are_all_pr_repos_public(){
+    REPO_IS_PRIVATE="true"
+    printf '\n%s\n' "Testing if repos specified in pr_branches are all public" >&2
+    PR_BRANCHES="$@"
+    for i in $PR_BRANCHES; do
+      echo "Testing URL: ${i}" >&2
+      # Convert github URL to shadow-robot/repo_name
+      user_slash_repo=$(echo $i | sed -r 's/.*github\.com\///g' | sed -r s'/\/tree.*//g' | sed -r 's/\/pull.*//g')
+      REPO_IS_PUBLIC=$(is_repo_public $user_slash_repo)
+      if [[ $REPO_IS_PUBLIC == "false" ]]; then
+        REPO_IS_PRIVATE="false"
+        break
+      elif [[ $REPO_IS_PUBLIC == "403" ]]; then
+        REPO_IS_PRIVATE="403"
+        break
+      fi
+    done
+    echo $REPO_IS_PRIVATE
+}
+
+check_github_next_steps(){
+    PUBLIC_REPO_STATUS=$1
+    NEXT_STEP=0
+    if [[ $PUBLIC_REPO_STATUS == "403" ]]; then
+        printf '%s' "WARNING: Rate limit exceeded for github api requests." >&2
+        printf '%s\n' " It is not currently possible to confirm whether all the URLs specified in PR_BRANCHES belong to public repos" >&2
+        printf '%s\n' "(Rate limits only last for 60 minutes, if you are unsure then please try again later)" >&2
+        if [[ $(confirm "Would you like to create a key and authenticate it? y/N") == "y" ]]; then
+            NEXT_STEP="generate_key"
+        elif [[ $(confirm "Would you like to continue without this check? y/N") == "y" ]] ; then
```

Comment:
Is there a reason `y/N` is written as `y/N` and not `y/n` as in other parts of the script (e.g. line `286`)?

Reply:
capital refers to default value, there are two different implementations of confirmation in this script, wasn't sure it was the best use of time to refactor and test the existing confirmation but if you think that's best then you'd probably be correct so I'd be happy to

Reply:
Thanks for clarifying, wasn't aware that was a convention. I agree that it's probably not worth the time commitment to refactor and test the existing confirmations.